### PR TITLE
Remove my `scalar_copy_backend_type` optimization attempt

### PR DESCRIPTION
**Pull request by scottmcm**
I added this back in https://github.com/rust-lang/rust/pull/111999 , but I no longer think it's a good idea
- It had to get scaled back to only power-of-two things to not break a bunch of targets
- LLVM seems to be getting better at memcpy removal anyway
- Introducing vector instructions has seemed to sometimes (https://github.com/rust-lang/rust/pull/115515#issuecomment-1750069529) make autovectorization worse

So this removes it from the codegen crates entirely, and instead just tries to use <https://doc.rust-lang.org/nightly/nightly-rustc/rustc_codegen_ssa/traits/builder/trait.BuilderMethods.html#method.typed_place_copy> instead of direct `memcpy` so things will still use load/store when a type isn't `OperandValue::Ref`.

### CODE REVIEW

File: compiler/rustc_codegen_ssa/src/mir/operand.rs [hunk 1]
```diff
@@ -419,7 +418,14 @@ impl<'a, 'tcx, V: CodegenObject> OperandValue<V> {
                     bx.store_with_flags(val, dest.llval, dest.align, flags);
                     return;
                 }
-                base::memcpy_ty(bx, dest.llval, dest.align, r, source_align, dest.layout, flags)
+                bx.memcpy_known_size(
```
Comment (compiler-errors) [inline compiler/rustc_codegen_ssa/src/mir/operand.rs:421]: 🤔 why isn't this also just using `typed_place_copy`?

also, `memcpy_known_size` has one callsite (this one) -- you should inline it. i don't really see the value of exposing it if it's only being used once, seems to add more confusion to the api imo.

**Comment by fee1-dead** [comment]
r? compiler

**Comment by compiler-errors** [comment]
I already started this review anyways

r? compiler-errors

File: compiler/rustc_codegen_ssa/src/traits/builder.rs [hunk 2]
```diff
@@ -280,17 +280,31 @@ pub trait BuilderMethods<'a, 'tcx>:
...
             // If we're not optimizing, the aliasing information from `memcpy`
             // isn't useful, so just load-store the value for smaller code.
             let temp = self.load_operand(src);
-            temp.val.store(self, dst);
+            temp.val.store_with_flags(self, dst, flags);
+        } else if flags.contains(MemFlags::NONTEMPORAL) {
+            // HACK(nox): This is inefficient but there is no nontemporal memcpy.
+            let ty = self.backend_type(dst.layout);
+            let val = self.load(ty, src.llval, src.align);
+            self.store_with_flags(val, dst.llval, dst.align, flags);
```
Comment (compiler-errors) [inline compiler/rustc_codegen_ssa/src/traits/builder.rs:304]: What is the difference between this and the above code? It feels subtly different but only barely

Comment (compiler-errors) [reply compiler/rustc_codegen_ssa/src/traits/builder.rs:304]: nontemporal sounds scary enough that i feel like we should reorder these branches, then? like, check nontemporal first? unless you're particularly convinced there will be no side effects.

File: tests/codegen/issues/issue-122805.rs [hunk 3]
```diff
@@ -9,15 +14,27 @@
...
+// OPT2-NEXT: store i16
+// OPT2-NEXT: getelementptr inbounds i8, {{.+}} 14
+// OPT2-NEXT: store i16
+// OPT3LINX64: load <8 x i16>
+// OPT3LINX64-NEXT: call <8 x i16> @llvm.bswap
+// OPT3LINX64-NEXT: store <8 x i16>
+// OPT3WINX64: load <8 x i16>
+// OPT3WINX64-NEXT: call <8 x i16> @llvm.bswap
+// OPT3WINX64-NEXT: store <8 x i16>
 // CHECK-NEXT: ret void
```
[the remark this answers is not in the document]
Comment (DianQK) [reply tests/codegen/issues/issue-122805.rs:38]: Ah, thanks! I hadn't noticed that different optimization levels yield different optimization effects.
I believe it makes sense to differentiate between O2 and O3 here.

The changes to O2 seem fine. I'll check again on this later.

**Review by dianqk — COMMENTED** [review]
> LLVM seems to be getting better at memcpy removal anyway.

I think we can expect to see more significant changes after https://github.com/llvm/llvm-project/pull/87190.

File: tests/codegen/issues/issue-122805.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (DianQK) [reply tests/codegen/issues/issue-122805.rs:40]: I confirm that is fine. I'm just curious why the results are the same for O2 and O3 is in LLVM's opt. 
I've found that Rust's O2 disables LLVM's SLP vectorization pass:  https://github.com/rust-lang/rust/blob/f96442b4482fef6a50b65f2284bcfb8921320183/compiler/rustc_codegen_ssa/src/back/write.rs#L248-L249.

**Pull request merged**